### PR TITLE
Support raw kubeconfig input for attach clusters as an alternative input method

### LINKED ISSUE #77: kubeconfig input alternate options
### Is your feature request related to a problem? Please describe.

When using the attach_cluster_with_kubeconfig you must specify a path to the kubeconfig using the kubeconfig_file variable and can't use the content of an output or block as an input to the config needed.  There are extra modules needed to create a file in order to do this and creates more dependencies.

### Describe the solution you'd like

For example, when attaching an AKS cluster, it would be nice to use the [kube_config_raw](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/data-sources/kubernetes_cluster#kube_config_raw) or kube_config_admin_raw value as an input for the kubeconfig needed to attach a cluster to TMC.  Currently for an AKS attach, you must create the file through another method like local_file from the kube_config_raw variable and pass that file location to the TMC module.  So in the function of attach_k8s_cluster maybe have another input option that could be the content of another ouput.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Vasundharashukla**
1. **What this PR does / why we need it**:
Add kubeconfig as raw input to the terraform script as an alternative option along with passing the file address for attach cluster resource.

2. **Which issue(s) this PR fixes**

        (optional, in `fixes #<issue number>` format, will close the issue(s) when PR gets merged):

        Fixes #77 

3. **Cluster attach verification**

Screenshot 2023-02-21 at 4 24 15 PM

Screenshot 2023-02-21 at 4 25 26 PM

### CODE REVIEW

File: internal/resources/cluster/resource_cluster.go [hunk 1]
```diff
@@ -218,15 +227,27 @@ func resourceClusterCreate(ctx context.Context, d *schema.ResourceData, m interf
...
 	var (
 		kubeconfigfile string
+		rawkubeconfig  string
 		k8sclient      k8sClient.Client
 		err            error
 	)
 
 	if _, ok := d.GetOk(attachClusterKey); ok {
-		kubeconfigfile, _ = d.Get(helper.GetFirstElementOf(attachClusterKey, attachClusterKubeConfigKey)).(string)
+		if v, ok := d.GetOk(helper.GetFirstElementOf(attachClusterKey, attachClusterKubeConfigPathKey)); ok {
```
Comment (shreyassreenivas) [inline internal/resources/cluster/resource_cluster.go:236]: can you use switch here to construct the k8s client from file or raw config

File: internal/resources/cluster/resource_cluster.go [hunk 2]
```diff
@@ -250,10 +271,10 @@ func resourceClusterCreate(ctx context.Context, d *schema.ResourceData, m interf
 	// always run
 	d.SetId(clusterResponse.Cluster.Meta.UID)
 
-	if kubeconfigfile != "" {
+	if _, ok := d.GetOk(attachClusterKey); ok {
```
Comment (shreyassreenivas) [inline internal/resources/cluster/resource_cluster.go:274]: This condition functionality will fail when attachClusterKey is defined but none of the attributes are added inside the block

File: internal/resources/cluster/resource_cluster.go [hunk 3]
```diff
@@ -425,18 +446,71 @@ func resourceClusterInPlaceUpdate(ctx context.Context, d *schema.ResourceData, m
...
 }
 
-func getK8sClient(kubeConfigFile string) (k8sClient.Client, error) {
-	restConfig, err := clientcmd.BuildConfigFromFlags("", kubeConfigFile)
-	if err != nil {
-		return nil, errors.WithMessagef(err, "Invalid kubeconfig file path provided, filepath : %s", kubeConfigFile)
+type (
+	KubeConfigOption func(*KubeConfig)
+
+	KubeConfigSpec struct {
```
Comment (shreyassreenivas) [inline internal/resources/cluster/resource_cluster.go:452]: This change should be in next MR and it is not related to the present change

File: internal/resources/cluster/resource_cluster.go [hunk 4]
```diff
@@ -425,18 +446,71 @@ func resourceClusterInPlaceUpdate(ctx context.Context, d *schema.ResourceData, m
...
+	}
+
+	KubeConfig struct {
+		Path string
+		Raw  string
+		Spec *KubeConfigSpec
+	}
+)
+
+func WithPath(p string) KubeConfigOption {
```
Comment (shreyassreenivas) [inline internal/resources/cluster/resource_cluster.go:466]: Good use of functional arguments

File: internal/resources/cluster/resource_cluster.go [hunk 5]
```diff
@@ -425,18 +446,71 @@ func resourceClusterInPlaceUpdate(ctx context.Context, d *schema.ResourceData, m
...
+	for _, o := range opts {
+		o(cfg)
+	}
+
+	var (
+		restConfig *rest.Config
+		err        error
+	)
+
+	if cfg.Path != "" {
```
Comment (shreyassreenivas) [inline internal/resources/cluster/resource_cluster.go:490]: use switch statement here

File: internal/resources/cluster/resource_cluster.go (hunk 2 above)
Comment (Vasundharashukla) [reply internal/resources/cluster/resource_cluster.go:274]: For that terraform will error out from schema as expected min items is 1.

Comment (shreyassreenivas) [reply internal/resources/cluster/resource_cluster.go:274]: That block has description , so if you just provide description and not other values how will ur code work ?

Comment (Vasundharashukla) [reply internal/resources/cluster/resource_cluster.go:274]: Is the description really needed as a parameter to be passed, I just see it defined in the schema and not used any where else in the code base, is it something that will be largely used by the provider users?

Comment (shreyassreenivas) [reply internal/resources/cluster/resource_cluster.go:274]: I am not talking about the use case here.
Pointing to the the code and the corner case which you are missing

Comment (shreyassreenivas) [reply internal/resources/cluster/resource_cluster.go:274]: Description inside the attach k8s might be used by the user to describe something which would help them identify the kubeconfig which is being used.

File: internal/resources/cluster/resource_cluster.go (hunk 1 above)
Comment (Vasundharashukla) [reply internal/resources/cluster/resource_cluster.go]: Done

File: internal/resources/cluster/resource_cluster.go (hunk 2 above)
Comment (Vasundharashukla) [reply internal/resources/cluster/resource_cluster.go:328]: Added schema validation which will ensure that one of the ways of defining kubeconfig have been added and it will throw validation error from terraform to handle this case.

File: internal/resources/cluster/resource_cluster.go (hunk 5 above)
Comment (Vasundharashukla) [reply internal/resources/cluster/resource_cluster.go]: Done

File: docs/data-sources/cluster.md [hunk 6]
```diff
@@ -80,7 +80,7 @@ data "tanzu-mission-control_cluster" "read_tkg_aws_cluster" {
 
 ### Optional
 
-- `attach_k8s_cluster` (Block List, Max: 1) (see [below for nested schema](#nestedblock--attach_k8s_cluster))
+- `attach_k8s_cluster` (Block List, Max: 2) (see [below for nested schema](#nestedblock--attach_k8s_cluster))
```
Comment (shreyassreenivas) [inline docs/data-sources/cluster.md:83]: why is this max two ?

File: internal/resources/cluster/resource_cluster.go [hunk 7]
```diff
@@ -27,6 +31,12 @@ import (
 	"github.com/vmware/terraform-provider-tanzu-mission-control/internal/resources/common"
 )
 
+const (
+	unknown = iota
```
Comment (shreyassreenivas) [inline internal/resources/cluster/resource_cluster.go:35]: is this really needed ? can this be avoided ?

File: internal/resources/cluster/resource_cluster.go [hunk 8]
```diff
@@ -39,6 +49,7 @@ func ResourceTMCCluster() *schema.Resource {
 		UpdateContext: resourceClusterInPlaceUpdate,
 		DeleteContext: resourceClusterDelete,
 		Schema:        clusterSchema,
+		CustomizeDiff: ValidateKubeConfig,
```
Comment (shreyassreenivas) [inline internal/resources/cluster/resource_cluster.go:52]: adding schema validation in customiseDiff is wrong way. 
From what I understand, customiseDiff is used to show the diffs and override the diffs shown or not.
Can you check the function definition in framework ? I might be wrong here

File: internal/resources/cluster/resource_cluster.go [hunk 9]
```diff
@@ -94,26 +105,40 @@ func constructFullname(d *schema.ResourceData) (fullname *clustermodel.VmwareTan
...
-			attachClusterDescriptionKey: {
-				Type:        schema.TypeString,
-				Description: "Attach cluster description",
-				Optional:    true,
+var (
+	attachCluster = &schema.Schema{
+		Type:     schema.TypeList,
+		Optional: true,
+		MinItems: 1,
+		MaxItems: 2,
```
Comment (shreyassreenivas) [inline internal/resources/cluster/resource_cluster.go:113]: why should the max items be 2

File: internal/resources/cluster/resource_cluster.go [hunk 10]
```diff
@@ -213,20 +238,70 @@ func flattenSpec(spec *clustermodel.VmwareTanzuManageV1alpha1ClusterSpec) (data
 	return []interface{}{flattenSpecData}
 }
 
+func ValidateKubeConfig(ctx context.Context, diff *schema.ResourceDiff, i interface{}) error {
```
Comment (shreyassreenivas) [inline internal/resources/cluster/resource_cluster.go:241]: The kubeconfig can be validated by checking if the k8s client is nil or not

File: internal/resources/cluster/resource_cluster.go (hunk 2 above)
Comment (shreyassreenivas) [reply internal/resources/cluster/resource_cluster.go:328]: Here, we should be checking is k8s client is nil or not

File: docs/data-sources/cluster.md (hunk 6 above)
Comment (Vasundharashukla) [reply docs/data-sources/cluster.md:83]: One element is description, one is the kube_config input (either of the 2 available inputs), so that user doesn't provide multiple inputs.

File: internal/resources/cluster/resource_cluster.go (hunk 7 above)
Comment (Vasundharashukla) [reply internal/resources/cluster/resource_cluster.go:35]: For the switch case you asked to use instead of if conditions.

File: internal/resources/cluster/resource_cluster.go (hunk 9 above)
Comment (Vasundharashukla) [reply internal/resources/cluster/resource_cluster.go:113]: Same reason as above

File: internal/resources/cluster/resource_cluster.go (hunk 10 above)
Comment (Vasundharashukla) [reply internal/resources/cluster/resource_cluster.go:241]: A schema validation checks before even the call is made to TMC, so one the first layer we get the response.

File: docs/data-sources/cluster.md (hunk 6 above)
Comment (shreyassreenivas) [reply docs/data-sources/cluster.md:83]: this limit is for the entries in they list. 
kuebconfig and description together will be on entry.

File: internal/resources/cluster/resource_cluster.go (hunk 7 above)
Comment (shreyassreenivas) [reply internal/resources/cluster/resource_cluster.go:35]: Switch can be implemented without this iota and this is not needed

File: internal/resources/cluster/resource_cluster.go (hunk 10 above)
Comment (shreyassreenivas) [reply internal/resources/cluster/resource_cluster.go:241]: you can use : ValidateFunc or ExactlyOneOf  []string attributes from the schema. 
We should not use diff customise diff to handle schema validations

File: internal/resources/cluster/resource_cluster.go (hunk 7 above)
Comment (Vasundharashukla) [reply internal/resources/cluster/resource_cluster.go:35]: Can you suggest the switching parameter passing the input kubeconfig value to k8sclient?

File: internal/resources/cluster/resource_cluster.go (hunk 10 above)
Comment (Vasundharashukla) [reply internal/resources/cluster/resource_cluster.go:241]: I believe than we need to change the same for other resources as well, customise diff has been used in other resources as well

File: internal/resources/cluster/resource_cluster.go (hunk 7 above)
Comment (shreyassreenivas) [reply internal/resources/cluster/resource_cluster.go:35]: you can use 
switch true{

}


or just switch {

}

This will solve ur use case

File: internal/resources/cluster/resource_cluster.go (hunk 10 above)
Comment (shreyassreenivas) [reply internal/resources/cluster/resource_cluster.go:241]: if it is only limited to schema validation and something which can be handled via validateFunc we should be using it. 
if other resources needs change, create a task and it can be picked in the future

File: internal/resources/cluster/resource_cluster.go (hunk 8 above)
Comment (Vasundharashukla) [reply internal/resources/cluster/resource_cluster.go:52]: Ref- https://github.com/hashicorp/terraform/pull/14887#issue-231839744

Usecase of `CustomizeDiff` from the terraform org: 

> Vetoing changes to the diff that could not be reasonably carried out (ie: illegal AWS RDS operations) by kicking back an error to the CustomizeDiff function, much like the other CRUD functions.

File: internal/resources/cluster/resource_cluster.go (hunk 2 above)
Comment (Vasundharashukla) [reply internal/resources/cluster/resource_cluster.go:336]: Added the check

File: docs/data-sources/cluster.md (hunk 6 above)
Comment (Vasundharashukla) [reply docs/data-sources/cluster.md]: Done

File: internal/resources/cluster/resource_cluster.go (hunk 7 above)
Comment (Vasundharashukla) [reply internal/resources/cluster/resource_cluster.go]: Removed it and added switch case

File: internal/resources/cluster/resource_cluster.go (hunk 8 above)
Comment (Vasundharashukla) [reply internal/resources/cluster/resource_cluster.go:46]: As per the above use case, we can use `CustomizeDiff` for validations.

File: internal/resources/cluster/resource_cluster.go (hunk 10 above)
Comment (Vasundharashukla) [reply internal/resources/cluster/resource_cluster.go:235]: `validateFunc` is being used to check non empty Kube_config inputs. Here we are checking some extra corner cases such as not giving more than one Kube_config input type

**Comment by ishangupta-ds** [comment]
Did we add the check for 'ready_wait_timeout', this should be optional if the customer does not want to attach the cluster right?

Cc: @shreyassreenivas

File: docs/data-sources/cluster.md [hunk 11]
```diff
@@ -98,7 +98,8 @@ data "tanzu-mission-control_cluster" "read_tkg_aws_cluster" {
 Optional:
 
 - `description` (String) Attach cluster description
-- `kubeconfig_file` (String) Attach cluster KUBECONFIG path
+- `kube_config_path` (String) Attach cluster KUBECONFIG path
```
[the remark this answers is not in the document]
Comment (Vasundharashukla) [reply docs/data-sources/cluster.md]: Reversed the change

File: internal/resources/cluster/resource_cluster.go [hunk 12]
```diff
@@ -213,20 +232,84 @@ func flattenSpec(spec *clustermodel.VmwareTanzuManageV1alpha1ClusterSpec) (data
...
+		}
+
+		isKubeConfigPresent := func(typeKey string) bool {
+			if value, ok := d.GetOk(helper.GetFirstElementOf(attachClusterKey, typeKey)); ok {
+				if value != nil {
+					kubeConfig = value
+					return true
+				}
+
+				return false
```
Comment (shreyassreenivas) [inline internal/resources/cluster/resource_cluster.go:293]: This return is duplicate

File: internal/resources/cluster/resource_cluster.go [hunk 13]
```diff
@@ -274,6 +357,66 @@ func resourceClusterCreate(ctx context.Context, d *schema.ResourceData, m interf
...
+		}
+	}
+
+	if restConfig == nil {
+		return nil, errors.WithMessagef(err, "Kubeconfig not provided.")
+	}
+
+	restConfig.Timeout = 10 * time.Second
+
+	client, err := k8sClient.New(restConfig, k8sClient.Options{})
```
Comment (shreyassreenivas) [inline internal/resources/cluster/resource_cluster.go:412]: check what is the value of the client in the case of the error and you can use it the above if condition check

**Review by shreyassreenivas — CHANGES_REQUESTED** [review]

File: internal/resources/cluster/resource_cluster.go (hunk 2 above)
Comment (Vasundharashukla) [reply internal/resources/cluster/resource_cluster.go:335]: Added

**Review by ishangupta-ds — DISMISSED** [review]

File: internal/resources/cluster/resource_cluster.go [hunk 14]
```diff
@@ -94,26 +99,40 @@ func constructFullname(d *schema.ResourceData) (fullname *clustermodel.VmwareTan
...
+					Optional:     true,
+					ForceNew:     true,
+					Sensitive:    true,
+					ValidateFunc: validation.StringIsNotWhiteSpace,
+				},
+				attachClusterDescriptionKey: {
+					Type:         schema.TypeString,
+					Description:  "Attach cluster description",
+					Optional:     true,
+					ValidateFunc: validation.StringIsNotWhiteSpace,
```
[the remark this answers is not in the document]
Comment (Vasundharashukla) [reply internal/resources/cluster/resource_cluster.go:129]: Yes, tested. Refer to the comment [here](https://github.com/vmware/terraform-provider-tanzu-mission-control/pull/125#discussion_r1121620614)

File: internal/resources/cluster/resource_cluster.go [hunk 15]
```diff
@@ -274,6 +357,66 @@ func resourceClusterCreate(ctx context.Context, d *schema.ResourceData, m interf
 	return append(diags, dataSourceTMCClusterRead(context.WithValue(ctx, contextMethodKey{}, "create"), d, m)...)
 }
 
+type (
+	KubeConfigOption func(*KubeConfig)
```
[the remark this answers is not in the document]
Comment (Vasundharashukla) [reply internal/resources/cluster/resource_cluster.go]: made all variables local

File: internal/resources/cluster/resource_cluster.go [hunk 16]
```diff
@@ -274,6 +357,66 @@ func resourceClusterCreate(ctx context.Context, d *schema.ResourceData, m interf
...
+
+	KubeConfig struct {
+		Path string
```
[the remark this answers is not in the document]
Comment (Vasundharashukla) [reply internal/resources/cluster/resource_cluster.go]: done

File: internal/resources/cluster/resource_cluster.go (hunk 12 above)
Comment (Vasundharashukla) [reply internal/resources/cluster/resource_cluster.go]: removed

File: internal/resources/cluster/resource_cluster.go (hunk 13 above)
Comment (Vasundharashukla) [reply internal/resources/cluster/resource_cluster.go:411]: added the check above

File: docs/resources/cluster.md [hunk 17]
```diff
@@ -42,6 +42,7 @@ resource "tanzu-mission-control_cluster" "attach_cluster_without_apply" {
     cluster_group = "default" # Default: default
   }
 
+  ready_wait_timeout = "0s"
```
Comment (shreyassreenivas) [inline docs/resources/cluster.md:45]: is this intended change ?

File: docs/resources/cluster.md [hunk 18]
```diff
@@ -1029,6 +1061,11 @@ Optional:
 - `disk_size` (String) Root disk size in gigabytes for the VM
 - `memory` (String) Memory associated with the node in megabytes
 
+
```
Comment (shreyassreenivas) [inline docs/resources/cluster.md:1064]: why white spaces ?

File: examples/resources/cluster/resource_attach_cluster.tf [hunk 19]
```diff
@@ -13,5 +13,6 @@ resource "tanzu-mission-control_cluster" "attach_cluster_without_apply" {
     cluster_group = "default" # Default: default
   }
 
+  ready_wait_timeout = "0s"
```
Comment (shreyassreenivas) [inline examples/resources/cluster/resource_attach_cluster.tf:16]: can you explain why is this needed ?

File: internal/resources/cluster/resource_cluster.go [hunk 20]
```diff
@@ -213,24 +232,87 @@ func flattenSpec(spec *clustermodel.VmwareTanzuManageV1alpha1ClusterSpec) (data
...
+			k8sclient, err = getK8sClient(withRaw(rawKubeConfig))
+		}
+
+		if err != nil {
+			log.Println("[ERROR] error while creating kubernetes client: ", err.Error())
+			return diag.FromErr(err)
+		}
+
+		if k8sclient == nil {
+			err = errors.New("k8s client not found")
```
Comment (shreyassreenivas) [inline internal/resources/cluster/resource_cluster.go:312]: check the error message, k8s client is something we are creating, so not found will be a wrong error

File: internal/resources/cluster/resource_cluster.go [hunk 21]
```diff
@@ -263,7 +345,7 @@ func resourceClusterCreate(ctx context.Context, d *schema.ResourceData, m interf
 
 		log.Printf("[INFO] Applying %s cluster's deployment link manifest objects on to kubernetes cluster", constructFullname(d).ToString())
 
-		err = manifest.Create(k8sclient, deploymentManifests, true)
+		err = manifest.Create(*k8sclient, deploymentManifests, true)
```
Comment (shreyassreenivas) [inline internal/resources/cluster/resource_cluster.go:348]: modify the manifest.Create function parameter to accept pointer k8sclient

File: examples/resources/cluster/resource_attach_cluster.tf (hunk 19 above)
Comment (ishangupta-ds) [reply examples/resources/cluster/resource_attach_cluster.tf:16]: @shreyassreenivas I asked for this change, basically if we are not attaching the cluster and only creating it (without the attach block) then I don't see any reason to block the terraform apply for 3 minutes.

WDYT

File: docs/resources/cluster.md (hunk 17 above)
Comment (ishangupta-ds) [reply docs/resources/cluster.md:45]: Yes

File: docs/data-sources/cluster.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (Vasundharashukla) [reply docs/data-sources/cluster.md]: that's autogenerated from the tool, not something added by the developer.

File: docs/data-sources/cluster.md [hunk 22]
```diff
@@ -99,6 +99,7 @@ Optional:
 
 - `description` (String) Attach cluster description
 - `kubeconfig_file` (String) Attach cluster KUBECONFIG path
+- `kube_config_raw` (String, Sensitive) Attach cluster KUBECONFIG
```
[the remark this answers is not in the document]
Comment (Vasundharashukla) [reply docs/data-sources/cluster.md]: yes, that's why for cluster scenario without kubeconfig, it doesn't make sense to wait for even 3 minutes. So, documentation can support this case.

File: examples/resources/cluster/resource_attach_cluster.tf (hunk 19 above)
Comment (Vasundharashukla) [reply examples/resources/cluster/resource_attach_cluster.tf:16]: Added to the documentation, so that users can take reference and not wait for 3m when kubeconfig is not present.

File: internal/resources/cluster/resource_cluster.go (hunk 20 above)
Comment (Vasundharashukla) [reply internal/resources/cluster/resource_cluster.go]: changed

File: internal/resources/cluster/resource_cluster.go (hunk 21 above)
Comment (Vasundharashukla) [reply internal/resources/cluster/resource_cluster.go]: done

File: internal/resources/cluster/resource_cluster.go [hunk 23]
```diff
@@ -274,6 +356,66 @@ func resourceClusterCreate(ctx context.Context, d *schema.ResourceData, m interf
...
+	}
+}
+
+func withRaw(r string) kubeConfigOption {
+	return func(config *kubeConfig) {
+		config.raw = r
+	}
+}
+
+func getK8sClient(opts ...kubeConfigOption) (*k8sClient.Client, error) {
```
Comment (shreyassreenivas) [inline internal/resources/cluster/resource_cluster.go:380]: would it make sense to move this function inside the package where the k8s client is being consumed ie manifest package.

If you are moving, move both the function and the type variables

File: internal/resources/cluster/manifest/manifest.go [hunk 24]
```diff
@@ -31,7 +31,7 @@ func Create(
 		return errors.WithMessage(err, "failure to fetch attach manifests")
 	}
 
-	toBeCleaned, err := objectsToBeCleaned(k8sclient, manifests, forceClean)
+	toBeCleaned, err := objectsToBeCleaned(*k8sclient, manifests, forceClean)
```
Comment (shreyassreenivas) [inline internal/resources/cluster/manifest/manifest.go:34]: objectsToBeCleaned(*k8sclient, manifests, forceClean)

*k8sclient, looks odd and should be avoided, so can you change the function definition for objectsToBeCleaned and 
createObjects

File: internal/resources/cluster/resource_cluster.go (hunk 23 above)
Comment (Vasundharashukla) [reply internal/resources/cluster/resource_cluster.go:380]: Then we would need to export all variables and functions by making them global. Since, only cluster resource is using this, we can have it here.

**Review by ishangupta-ds — APPROVED** [review]

**Review by shreyassreenivas — APPROVED** [review]
